clang/lib/AST/DeclBase.cpp

```diff
@@ -1474,6 +1474,11 @@ DeclContext *DeclContext::getPrimaryContext() {
   case Decl::ObjCCategoryImpl:
     return this;
 
+  case Decl::CXXRecord:
+    if (auto *OPD = dyn_cast<CXXRecordDecl>(this))
+      if (auto *Def = OPD->getDefinition())
+        return Def;
+    return this;
   default:
```

Comment:
Why is this necessary?

Reply:
This is required to fix for these tests to pass:
```
  Clang :: AST/HLSL/ByteAddressBuffer-AST.hlsl
  Clang :: AST/HLSL/OutArgExpr.hlsl
  Clang :: AST/HLSL/pch.hlsl
  Clang :: AST/HLSL/pch_hlsl_buffer.hlsl
  Clang :: AST/HLSL/pch_with_buf.hlsl
  Clang :: CodeGenHLSL/builtins/ByteAddressBuffers-constructors.hlsl
```

Those test fail with the error message below if such change is not present:
```
llvm/include/llvm/Support/Casting.h:662: decltype(auto) llvm::dyn_cast(From *) [To = clang::TagType, From = const clang::Type]: Assertion `detail::isPresent(Val) && "dyn_cast on a non-existent value"' failed.
```

Reply:
I am not 100% sure, but my understanding is, `ByteAddressBuffer` is the first buffer to not use `addSimpleTemplateParams`. Such method, calls `Builder.addTypeParameter`, which modifies the declaration. 

@bogner  @llvm-beanz, do any of you can correct/confirm this fix, please?

Reply:
Ok, that explains why the *pch* tests are now failing - they include the new buffer types.

I believe the default case in `DeclContext::getPrimaryContext` should be covering the CXXRecordDecl case as well. Since it hits the assert, my guess is that the CXXRecordDecl is probably not set up as expected. Maybe the definition is not complete? What is the difference between the records that use `addSimpleTemplateParams` and this one that does not when executing `DeclContext::getPrimaryContext`?

Reply:
The issue I was facing in the default case of `DeclContext::getPrimaryContext` was when calling `Tag->getTypeForDecl()` (line 1491). For `ByteAdrressBuffer` such method was returning `nullptr`.

For example, when calling `getTypeForDecl` for `StructuredBuffer` I get the response below:
```
InjectedClassNameType 0x5555668d3e20 'StructuredBuffer<element_type>' dependent
`-CXXRecord 0x5555668d3a60 'StructuredBuffer'
```

In contrast, if I call the same method for `ByteAddressBuffer` I get:
```
<<<NULL>>>
```

Reply:
Right, because the `TypeForDecl` is not set for `ByteAddressBuffer`. We need to make sure it gets set, otherwise the decl is not valid and it is going to bite us again someplace else.

We are creating the record in `BuiltinTypeDeclBuilder` constructor with `DelayTypeCreation` flag. For template classes the type gets created in `finalizeTemplateArgs` in the `ASTContext::getInjectedClassNameType` call, but that does not happed for non-template classes. We need to make sure it gets created and assigned, for example by adding `ASTContext::getTypeDeclType` call to the `BuiltinTypeDeclBuilder` destructor:

```
  ~BuiltinTypeDeclBuilder() {
    if (!Record->getTypeForDecl())
      Record->getASTContext().getTypeDeclType(Record,
                                              Record->getPreviousDecl());
    ... rest of the destructor code ....
  }

```